### PR TITLE
Introduce pre-order / post-order visitor pattern

### PR DESCRIPTION
**Pull request by TristonianJones**
Introduce a visitor with pre and post-order traversal methods.

This change also ensures that `MatchDescendents()` is bottom-up traversal
and that `Children()` can be used to model top-down traversal.

There are a couple of missing LICENSE notices added as well.

### CODE REVIEW

File: common/ast/navigable_test.go [hunk 1]
```diff
@@ -120,6 +135,44 @@ func TestNavigateAST(t *testing.T) {
 	}
 }
 
+func TestNavigableExprChildren(t *testing.T) {
```
Comment (l46kok) [inline common/ast/navigable_test.go]: Could we also test for post order traversal?

File: common/ast/navigable.go [hunk 2]
```diff
@@ -87,34 +101,146 @@ func AllMatcher() ExprMatcher {
...
+
+type visitOrder int
+
+const (
+	preOrder = iota + 1
+	postOrder
+)
+
+func visit(expr Expr, visitor Visitor, order visitOrder, depth, maxDepth int) {
+	if maxDepth > 0 && depth == maxDepth {
```
Comment (l46kok) [inline common/ast/navigable.go:195]: Thoughts on aborting on abnormally deep AST? [Something similar to this](https://github.com/google/cel-java/commit/26545aa8935427be712bafd3f0abb378f367f2f6#diff-ffb0c1526486c5e8de07dbaba6750bf21a7c2cf22b8bc8ebfba30f40005e88dcR95)

File: common/ast/navigable_test.go (hunk 1 above)
Comment (TristonianJones) [reply common/ast/navigable_test.go]: The post order ends up getting tested indirectly via `MaxID`, but I can add an explicit test to show it within this one.

Comment (l46kok) [reply common/ast/navigable_test.go]: The method is called in MaxID but I didn't come across any tests which validate the actual visiting order. Regardless, I think an explicit test is valuable.

File: common/ast/navigable.go (hunk 2 above)
Comment (TristonianJones) [reply common/ast/navigable.go:195]: I think I might want to set this as an option on the visitor. Java handles exceptions well, but go doesn't like panics, and I'm not sure I want to put an error return status on the visit. Let me think about this more and come back to it.

File: common/ast/navigable_test.go (hunk 1 above)
Comment (TristonianJones) [reply common/ast/navigable_test.go]: Agreed. I chose to show pre and post explicitly with an example pre order using `Children` as well.

File: common/ast/navigable.go (hunk 2 above)
Comment (TristonianJones) [reply common/ast/navigable.go:195]: I've added a TODO. We'll finalize the behavior before the next release tag.

**Review by l46kok — APPROVED** [review]

**Pull request merged**